clients/src/main/java/org/apache/kafka/common/utils/ByteBufferOutputStream.java

```diff
@@ -105,7 +105,7 @@ public int initialCapacity() {
         return initialCapacity;
     }
 
-    private void maybeExpandBuffer(int remainingRequired) {
+    public void maybeExpandBuffer(int remainingRequired) {
```

Comment:
My point is that `DefaultRecordBatch.writeHeader` was only called in `MemoryRecordsBuilder.writeDefaultBatchHeader` before this PR, and in `MemoryRecordsBuilder.writeDefaultBatchHeader` we only get `buffer` to book-keep its current position; we can modify the interface of `writeHeader` (and of course `writeEmptyHeader` as well) to pass in the `bufferStream` directly, then inside `writeHeader` we only call the api functions of the `bufferStream` and hence can remove this call before the `maybeExpandBuffer` call and still make this function internally?